branch-3.0 cherry-pick [fix](variant) fix the reading core caused by inserting nested column and scalar column in variant sub-column

amorynan opened the pull request:
### What problem does this PR solve?
backport: https://github.com/apache/doris/pull/53083
Issue Number: close #xxx

Related PR: #xxx

Problem Summary:

### Release note

None

### Check List (For Author)

- Test <!-- At least one of them must be included. -->
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason <!-- Add your reason?  -->

- Behavior changed:
    - [ ] No.
    - [ ] Yes. <!-- Explain the behavior change -->

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. <!-- Add document PR link here. eg: https://github.com/apache/doris-website/pull/1214 -->

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label <!-- Add branch pick label that this PR should merge into -->

Commit:
[fix](variant) fix the reading core caused by inserting nested column and scalar column in variant sub-column (#53083)

this pr main fix the problem which if we create table with
```variant_enable_flatten_nested```
then insert variant data:
```'{"nested":{"a":"1"}}``` and ```'{"nested":[{"a":1,"c":1.1},{"b":"1"}]}'```
we will meet core for reading
so we should forbid this table property
and for old data we insert different structure data will meet some error like this:
```
mysql> insert into vs values (2, '{"nested":{"a":"1"}}');
Query OK, 1 row affected (0.22 sec)
{'label':'label_165a8209698c4391_988c6532615017c4', 'status':'VISIBLE',
'txnId':'1011'}

mysql> insert into vs values (1,
'{"nested":[{"a":1,"c":1.1},{"b":"1"}]}');
ERROR 1105 (HY000): errCode = 2, detailMessage =
(10.16.10.6)[INTERNAL_ERROR]tablet 1752145213719 failed on majority
backends: [DATA_QUALITY_ERROR]PStatus:
(10.16.10.6)[DATA_QUALITY_ERROR]Ambiguous paths: v.nested.a vs
v.nested.a with different nested part true vs false
```

Thearas commented:
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://cwiki.apache.org/confluence/display/DORIS/How+to+process+your+PR).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

amorynan commented:
run buildall

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/8015b1c364411e90190002f6812d05738a4bfc93_8015b1c364411e90190002f6812d05738a4bfc93_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/8015b1c364411e90190002f6812d05738a4bfc93_8015b1c364411e90190002f6812d05738a4bfc93_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 80.98% (1171/1446) |
| Line Coverage     | 65.01% (19953/30690) |
| Region Coverage   | 66.58% (10061/15112) |
| Branch Coverage   | 56.19% (5308/9446) |

Commit:
[fix](variant) fix batch insert into with structure-conflicts strings (#53923)

fix batch insert into with structure-conflicts strings
Before this we were able to successfully insert, but will meet query
error
like this:
```
mysql> insert into var_nested_load_conflict values (3, '{"nested": [{"a": 2.5, "b": "123.1"}]}'),  (4, '{"nested": {"a": 2.5, "b": "123.1"}}');
Query OK, 2 rows affected (0.16 sec)
{'label':'label_9279242ae3fd40e2_aabe077db2d37bb9', 'status':'VISIBLE', 'txnId':'16028'}

mysql> desc var_nested_load_conflict;
+------------+---------------+------+-------+---------+-------+
| Field      | Type          | Null | Key   | Default | Extra |
+------------+---------------+------+-------+---------+-------+
| k          | bigint        | Yes  | true  | NULL    |       |
| v          | variant       | Yes  | false | NULL    | NONE  |
| v.nested.a | json          | Yes  | false | NULL    | NONE  |
| v.nested.b | json          | Yes  | false | NULL    | NONE  |
| v.nested.c | array<double> | Yes  | false | NULL    | NONE  |
+------------+---------------+------+-------+---------+-------+
5 rows in set (0.10 sec)

mysql> select * from var_nested_load_conflict;
ERROR 1105 (HY000): errCode = 2, detailMessage = (10.16.10.6)[INTERNAL_ERROR]Meet none array column when flatten nested array, path nested.b, type Nullable(JSONB)
```

So we don't allow this kind of insertion
```
mysql> insert into var_nested_load_conflict values (3, '{"nested": [{"a": 2.5, "b": "123.1"}]}'),  (4, '{"nested": {"a": 2.5, "b": "123.1"}}');
ERROR 1105 (HY000): errCode = 2, detailMessage = (10.16.10.6)[DATA_QUALITY_ERROR][E46] Ambiguous paths: nested.b vs nested.b with different nested part false vs true
```

Commit:
fix

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/ea10940351f265ec82a0bbf130c9a78e0ac50820_ea10940351f265ec82a0bbf130c9a78e0ac50820_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/ea10940351f265ec82a0bbf130c9a78e0ac50820_ea10940351f265ec82a0bbf130c9a78e0ac50820_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 80.98% (1171/1446) |
| Line Coverage     | 65.00% (19948/30690) |
| Region Coverage   | 66.56% (10059/15112) |
| Branch Coverage   | 56.17% (5306/9446) |

Commit:
fix code

amorynan commented:
run buildall

amorynan commented:
run p0

amorynan commented:
run cloud_p0

amorynan commented:
run external

Commit:
Merge branch 'branch-3.0' into pick_53083_to_doris_branch-3.0

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `62.50% (75/120)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/b4252bf711e494317d38c65ee76b8e38faac7bbd_b4252bf711e494317d38c65ee76b8e38faac7bbd/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/b4252bf711e494317d38c65ee76b8e38faac7bbd_b4252bf711e494317d38c65ee76b8e38faac7bbd/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 42.06% (11182/26583) |
| Line Coverage     | 32.57% (95856/294284) |
| Region Coverage   | 31.71% (49540/156223) |
| Branch Coverage   | 28.17% (25421/90236) |

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `83.33% (100/120)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/53897_b4252bf711e494317d38c65ee76b8e38faac7bbd_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/53897_b4252bf711e494317d38c65ee76b8e38faac7bbd_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 71.54% (18771/26237) |
| Line Coverage     | 63.94% (187925/293919) |
| Region Coverage   | 61.99% (112170/180935) |
| Branch Coverage   | 55.07% (56178/102020) |

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `83.33% (100/120)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/53897_b4252bf711e494317d38c65ee76b8e38faac7bbd_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/53897_b4252bf711e494317d38c65ee76b8e38faac7bbd_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 71.54% (18771/26237) |
| Line Coverage     | 63.93% (187911/293919) |
| Region Coverage   | 61.99% (112164/180935) |
| Branch Coverage   | 55.06% (56171/102020) |

Commit:
fix code

Commit:
fix fe format

amorynan commented:
run buildall

Commit:
fix fe format

amorynan commented:
run buildall

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `62.50% (75/120)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/7ffc567eb79ca32c22cd07c12a7ef4f8c44733dd_7ffc567eb79ca32c22cd07c12a7ef4f8c44733dd/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/7ffc567eb79ca32c22cd07c12a7ef4f8c44733dd_7ffc567eb79ca32c22cd07c12a7ef4f8c44733dd/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 42.07% (11184/26583) |
| Line Coverage     | 32.58% (95868/294284) |
| Region Coverage   | 31.72% (49548/156223) |
| Branch Coverage   | 28.18% (25424/90236) |

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `  25.00% (2/8)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/53897_7ffc567eb79ca32c22cd07c12a7ef4f8c44733dd/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/53897_7ffc567eb79ca32c22cd07c12a7ef4f8c44733dd/fe_report/index.html)

eldenmoon commented on fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java:
```diff
@@ -681,6 +681,10 @@ public class SessionVariable implements Serializable, Writable {
 
+    // disable variant flatten nested as session variable, default is true,
+    // which means disable variant flatten nested when create table
+    public static final String DISABLE_VARIANT_FLATTEN_NESTED = "disable_variant_flatten_nested";
```
rename to enable_xxx

eldenmoon reviewed (approved):
LGTM

dataroaring reviewed (approved):
LGTM
